src/ethereum/london/vm/gas.py

```diff
@@ -104,79 +137,55 @@ def calculate_memory_gas_cost(size_in_bytes: Uint) -> Uint:
 
 
 def calculate_gas_extend_memory(
-    memory: bytearray, start_position: U256, size: U256
-) -> Uint:
+    memory: bytearray, extensions: List[Tuple[U256, U256]]
+) -> ExtendMemory:
     """
     Calculates the gas amount to extend memory
 
     Parameters
     ----------
     memory :
         Memory contents of the EVM.
-    start_position :
-        Starting pointer to the memory.
-    size:
-        Amount of bytes by which the memory needs to be extended.
+    extensions:
+        List of extensions to be made to the memory.
+        Consists of a tuple of start position and size.
 
     Returns
     -------
-    to_be_paid : `ethereum.base_types.Uint`
-        returns `0` if size=0 or if the
-        size after extending memory is less than the size before extending
-        else it returns the amount that needs to be paid for extendinng memory.
-    """
-    if size == 0:
-        return Uint(0)
-    memory_size = Uint(len(memory))
-    before_size = ceil32(memory_size)
-    after_size = ceil32(Uint(start_position) + Uint(size))
-    if after_size <= before_size:
-        return Uint(0)
-    already_paid = calculate_memory_gas_cost(before_size)
-    total_cost = calculate_memory_gas_cost(after_size)
-    to_be_paid = total_cost - already_paid
-    return to_be_paid
-
-
-def calculate_call_gas_cost(
-    gas: Uint, gas_left: Uint, extra_gas: Uint
-) -> Uint:
+    extend_memory: `ExtendMemory`
     """
-    Calculates the gas amount for executing Opcodes `CALL` and `CALLCODE`.
-
-    Parameters
-    ----------
-    gas :
-        The amount of gas provided to the message-call.
-    gas_left :
-        The amount of gas left in the current frame.
-    extra_gas :
-        The amount of gas needed for transferring value + creating a new
-        account inside a message call.
+    size_to_extend = Uint(0)
+    to_be_paid = Uint(0)
+    current_size = Uint(len(memory))
+    for start_position, size in extensions:
+        if size == 0:
+            continue
+        before_size = ceil32(current_size)
+        after_size = ceil32(Uint(start_position) + Uint(size))
+        if after_size <= before_size:
+            continue
 
-    Returns
-    -------
-    call_gas_cost: `ethereum.base_types.Uint`
-        The total gas amount for executing Opcodes `CALL` and `CALLCODE`.
-    """
-    if gas_left < extra_gas:
-        raise OutOfGasError
+        size_to_extend += after_size - before_size
+        already_paid = calculate_memory_gas_cost(before_size)
+        total_cost = calculate_memory_gas_cost(after_size)
+        to_be_paid += total_cost - already_paid
 
-    gas = min(gas, max_message_call_gas(gas_left - extra_gas))
+        current_size = after_size
 
-    return gas + extra_gas
+    return ExtendMemory(to_be_paid, size_to_extend)
```

Comment:
Probably a dumb question, but this isn't equal to the cost of just expanding for the largest `after_size` out of all the `extensions`, is it?

Reply:
In most cases, yeah. But not always. If the `extension` with the largest `after_size` has a `size` of 0 (i.e. it has a very large `start_position`), the memory is not to be extended.